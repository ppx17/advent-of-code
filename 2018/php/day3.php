<?php
$data = $data ?? file_get_contents("../input/input-".basename(__FILE__, '.php').".txt");

preg_match_all(
    "/#(?<id>[0-9]+) @ (?<x>[0-9]+),(?<y>[0-9]+): (?<w>[0-9]+)x(?<h>[0-9]+)/",
    $data,
    $claims,
    PREG_SET_ORDER);

$len = count($claims);
for($i = 0; $i < $len; $i++) {
    $claims[$i]['x'] = (int)$claims[$i]['x'];
    $claims[$i]['y'] = (int)$claims[$i]['y'];
    $claims[$i]['w'] = (int)$claims[$i]['w'];
    $claims[$i]['h'] = (int)$claims[$i]['h'];
}

$grid = [];

foreach ($claims as $claim) {
    $maxX = ($claim['x'] + $claim['w']);
    $maxY = ($claim['y'] + $claim['h']);
    $startY = $claim['y'];
    for ($x = $claim['x']; $x < $maxX; $x++) {
        for ($y = $startY; $y < $maxY; $y++) {
            $idx = $x * 1000 + $y;
            if(isset($grid[$idx])) {
                $grid[$idx] = true;
            }else{
                $grid[$idx] = false;
            }
        }
    }
}

$count = 0;

foreach ($grid as $value) {
    if ($value === true) {
        $count++;
    }
}

echo "Part 1: " . $count . PHP_EOL;

function hasOverlap(array $grid, $claim) {
    for ($x = $claim['x']; $x < ($claim['x'] + $claim['w']); $x++) {
        for ($y = $claim['y']; $y < ($claim['y'] + $claim['h']); $y++) {
            $idx = $x * 1000 + $y;
            if($grid[$idx] === true) { return true; }
        }
    }
    return false;
}

foreach($claims as $claim) {
    if( ! hasOverlap($grid, $claim)) {
        echo "Part 2: " . $claim['id'] . PHP_EOL;
        break;
    }
}