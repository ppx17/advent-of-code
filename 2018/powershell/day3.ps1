Param(
    [string]$InputFile = '../input/input-day3.txt'
);

class Grid {
    [byte[]]$Grid;

    [int]$Width = 1000;
    [int]$Heigth = 1000;

    [byte]$Overlap = 2;

    Grid() {
        $this.Grid = [Byte[]]::new($this.Width*$this.Heigth);
    }

    [void]Claim([int]$x, [int]$y, [int]$w, [int]$h) {
        for($loopY = $y; $loopY -lt $y+$h; $loopY++) {
            for($loopX = $x; $loopX -lt $x+$w; $loopX++) {
                $idx = $loopX * $this.Width + $loopY;
                if($this.Grid[$idx] -lt $this.Overlap) {
                    $this.Grid[$idx]++;
                }
            }
        }
    }

    [bool]IsNotOverlapping([int]$x, [int]$y, [int]$w, [int]$h) {
        for($loopY = $y; $loopY -lt $y+$h; $loopY++) {
            for($loopX = $x; $loopX -lt $x+$w; $loopX++) {
                if($this.Grid[$loopX * $this.Width + $loopY] -eq $this.Overlap) {
                    return $false;
                }
            }
        }
        return $true;
    }

    [int]OverlapCount() {
        $count = 0;
        for($i=0;$i -lt $this.Grid.Length; $i++) {
            if($this.Grid[$i] -eq $this.Overlap) { $count++; }
        }
        return $count;
    }
}

$Claims = (Get-Content $InputFile) | Foreach-Object { $null = ($_ -match "#(?<id>[0-9]+) @ (?<x>[0-9]+),(?<y>[0-9]+): (?<w>[0-9]+)x(?<h>[0-9]+)"); return $Matches; };

$Grid = [Grid]::new();
foreach($Claim in $Claims) {
    $Grid.Claim($Claim['x'], $Claim['y'], $Claim['w'], $Claim['h']);
}

Write-Output ("Part 1: " + $Grid.OverlapCount());

foreach($Claim in $Claims) {
    if($Grid.IsNotOverlapping($Claim['x'], $Claim['y'], $Claim['w'], $Claim['h'])) {
        Write-Output ("Part 2: " + $Claim['id']);
        Exit;
    }
}