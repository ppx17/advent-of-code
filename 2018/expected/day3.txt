Part 1: 110383
Part 2: 129
